### DISCUSSION TITLE
RFC 30 Discussion

### DISCUSSION THREAD

**Discussion by Velfi** [smithy-lang/smithy-rs#2235] [General]
Hey @thomas-k-cameron. I'm creating this discussion so that we have a single place to talk about your PRs as a group.

I'm currently partway through merging your sub-PRs. Once those are all merged, we can talk next steps.

**Comment by thomas-k-cameron** [comment]
Alright, cool!
Thanks!

**Comment by thomas-k-cameron** [comment]
Just a heads up, we are updating the RFC30 concerning the handling of event_streams.
https://github.com/awslabs/smithy-rs/pull/2198#issue-1528489219

**Comment by Velfi** [comment]
@thomas-k-cameron I updated the `unstable-serde-support` branch to fix a few things I broke in the merges. Serialization for numbers isn't working but I think everything else is OK.

**Comment by thomas-k-cameron** [comment]
# TODO List

This is the best place to start. You can add this to codegen, even if it doesn't activate anything yet.

Link:
https://github.com/awslabs/smithy-rs/pull/1944#issuecomment-1371160038

- [x] Add a CargoDependency entry for serde
- [x] Create a SerdeDecorator class that extends ClientCodegenDecorator. This class should be declared in its own file, probably in the smithy-rs/codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations folder. When overriding the extras method, you get access to the RustCrate being generated and you can add features to it by calling rustCrate.mergeFeature(Feature("<your feature name>", default = true, listOf(<every CargoDependency needed by this feature>))). The decorator can add all features that you want to define.
- [ ] Write a test that ensures that the generated Cargo.toml contains the features declarations that you expect. The test would probably look similar to smithy-rs/codegen-core/src/test/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/CargoTomlGeneratorTest.kt and would be located near it too.

- [x] Implement human-readable + non-human-readable serialization for DateTime and Blob in aws_smithy_types

- [x] These two issues can be handled at the same time. You'd just submit an update to the aws-smithy-types runtime crate for this one I think. Extracting the declaration of Blob into its own module / folder would be a good idea. I don't want to add more to the lib.rs file if possible.
- [x] Implement Serialize and Deserialize for relevant data types in aws_smithy_types

- [x] Modify Kotlin's codegen so that generated Builder and non-Builder types implement Serialize and Deserialize

- [ ] Prepare examples

- [ ] Prepare reproducible compile time benchmark

**Comment by thomas-k-cameron** [comment]
I'm developing the reproducible compile time benchmark, but there are few things that I want to discuss with you guys.

Here's how the benchmark works.

- It's a CLI app
- It's based on AWS Batch 
- Configuration for compute environment/job queue/job definition... etc is written on TOML file and the app de-serialize it with the serde feature we implemented :>
- When you run the benchmark, it code-gens the sdk code with smithy-rs, and then runs cargo check for every each one of the generated SDK.

Advantage of this is, 
- With AWS Batch, you can always get a same, clean hardware so you will end up with the same outcome all the time
- This can become a real life showcase of the new feature. (finishes both example/benchmark)

Things that I haven't figured out is;

1. How do you want the benchmark result to be organized?

I'm thinking about creating a markdown table or something.  

2. How do you want to keep the result?

I think I'd love to invoke a github API or something so that it can be posted somewhere automatically.
(maybe you can make it a part of CI or something.)

I also thought about keeping it on a S3 bucket or something too; How do you want to proceed?  

3. Can I get the docker container image that you guys are using for the CI?

Where can I find the information about your docker container?  
It's not private right?

4. Which hardware should I use?

I'm thinking about `c7g.2xlarge` instance with 100GB of EBS.
c7 is new and it's based on AWS's own chip so I think it's great for what we are trying to do.
It also doesn't have any CPU burst like T4 series so the results should be more predictable.

You would need to change the instance type when it becomes unavailable in like 10 years or something, but I think we can figure it out when it happens.

**Reply by thomas-k-cameron** [reply]
Here is a draft pull request

https://github.com/awslabs/smithy-rs/pull/2309

**Reply by thomas-k-cameron** [reply]
@Velfi 
Hi,

can I have a comment on this please?
